Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift

```diff
@@ -293,6 +323,35 @@ extension SwiftPMWorkspace: SKCore.BuildSystem {
 
     return nil
   }
+
+  /// An event is relevant if it modifies a file that matches one of the file rules used by the SwiftPM workspace.
+  private func isFileEventRelevant(event: FileEvent) -> Bool {
+    // FIXME: fileRules should also include self.workspace.additionalFileRules and toolsVersion should be self.workspace.currentToolsVersion
+    let fileRules = FileRuleDescription.builtinRules
+    let toolsVersion = ToolsVersion.currentToolsVersion
+
+    guard let fileURL = event.uri.fileURL else {
+      return false
+    }
+    let filePath = AbsolutePath(fileURL.path)
+    switch event.type {
+    case .created, .deleted:
+      return fileRules.contains(where: { $0.match(path: filePath, toolsVersion: toolsVersion) })
+    case .changed:
+      return false
+    default: // Unknown file change type
+      return false
+    }
+  }
+
+  public func filesDidChange(_ events: [FileEvent]) {
+    if events.contains(where: { isFileEventRelevant(event: $0) }) {
+      // TODO: It should not be necessary to reload the entire package just to get build settings for one file.
+      orLog {
+        try self.reloadPackage()
```

Comment:
Do we need to do this on self.queue as well?

Reply:
It doesn’t need to be, because `BuildSystemManager` is already guarding against race conditions in `BuildSystemManager.filesDidChange` by informing the build systems about file changes serially on its own queue.

Reply:
I think it does since `registerForChangeNotifications` (on queue) could be called right before `filesDidChange`, and then `reloadPackage` (off queue) could mutate `packageGraph`/`fileToTarget`/`sourceDirToTarget ` while `registerForChangeNotifications` is reading it on the queue, right? So we need some way to synchronize that.

Reply:
Oh, yes, we also need to guard `packageGraph`, `fileToTarget` and `sourceDirToTarget` against race conditions. I only thought of `watchedFiles` and `delegate` before. Added the guards in https://github.com/apple/sourcekit-lsp/pull/443/commits/2e3b78312f40b89ba1c15f39a062c1881ad89e24.